extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -1423,14 +1424,18 @@ private void publishAllSegments(final Set<DataSegment> segments) throws IOExcept
               intervalsToDrop,
               destination.getReplaceTimeChunks(),
               task.getDataSource(),
-              destination.getSegmentGranularity()
+              destination.getSegmentGranularity(),
+              Limits.MAX_PARTITION_BUCKETS
           );
           segmentsWithTombstones.addAll(tombstones);
           numTombstones = tombstones.size();
         }
         catch (IllegalStateException e) {
           throw new MSQException(e, InsertLockPreemptedFault.instance());
         }
+        catch (IllegalArgumentException e) {
+          throw new MSQException(e, new TooManyBucketsFault(Limits.MAX_PARTITION_BUCKETS));
+        }
```

Comment:
I think this will also change, IAE doesn't seem valid here.

Reply:
Please see my related comment related to throwing MSQ faults from the indexing package. Also, this follow the same pattern as `InsertLockPreemptedFault` above

Reply:
We can pull out the exception above. 
`IAE` to me is more like entering a negative integer instead of a negative one, or a different type. The logic inside the helper operates before comparing it with the maxBuckets parameter. Or perhaps there are "XOR args" (enter exactly one) and the user passes both. 
As per InsertLockPreempted, it might be insufficiency at the time when it was written, but ISE seems more forgiving than IAE. Also, that logic in itself is also pretty fragile, if the logic switches to `DruidException`, then it would break.
At worst, IAE can have some special text inside, and we do a string compare before throwing this exception, however since we already have a special exception created, that seems like a valid path to follow

Reply:
My bad, you were referring to `TooManyBucketsException` and not the MSQ fault `TooManyBucketsFault`. Moved the exception to the `indexing-service` package for better exception handling semantics.